torch/utils/launch.py

```diff
@@ -0,0 +1,805 @@
+            begin = proc * cores_per_rank + ccl_worker_count
+            end = proc * cores_per_rank + cores_per_rank - 1
+            for i in range(begin, end + 1):
+                domain_binary |= (1 << i)
+            pin_domain += hex(domain_binary) + ","
+        pin_domain += "]"
+        return pin_domain
+
+    def get_ccl_worker_affinity(self, nproc_per_node, ccl_worker_count, total_cores):
+        '''
+        Computation and communication use different cores when using oneCCL
+        backend for distributed training. we use first ccl_worker_count cores of
+        every rank for ccl communication
+        '''
+        ppn = nproc_per_node
+        cores_per_rank = total_cores // ppn
+        affinity = ''
+        for proc in range(ppn):
+            for ccl_worker in range(ccl_worker_count):
+                affinity += str(proc * cores_per_rank + ccl_worker) + ","
+        affinity = affinity[:-1]
+        return affinity
+
+    def launch(self, args):
+        '''
+        Set ENVs and launch MPI process for distributed training.
+        '''
+        if args.nnodes > 1 and not os.path.exists(args.hostfile):
+            raise ValueError("hostfile is necessary when you use multi-node distributed training,"
+                             "Please create hostfile which include the ip list you used for distributed running")
+        elif args.nnodes > 1:
+            ipv4_addr_pattern = r"^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$"
+            ip_list = []
+            with open(args.hostfile) as f:
+                for line in f:
+                    line = line.strip().strip("\n")
+                    # is_valid = re.match(ipv4_addr_pattern, line)
+                    # if not is_valid:
+                    #     logger.error("{} is not valid IPV4 address".format(line))
+                    #     exit(-1)
+                    # else:
+                    #     ip_list.append(line)
+                    ip_list.append(line)
+            if len(ip_list) < args.nnodes:
+                logger.error("The number of IP {} should greater than nnodes parameters {}".format(len(ip_list), args.nnodes))
+                exit(-1)
+            master_check = False
+            dic = psutil.net_if_addrs()
+            for adapter in dic:
+                snicList = dic[adapter]
+                for snic in snicList:
+                    if snic.address == ip_list[0]:
+                        master_check = True
+            if not master_check:
+                logger.error("MASTER_ADDR is incorrect. Please make sure the first line {} in your hostfile is ip address of the current node".format(ip_list[0]))
+                exit(-1)
+
+            logger.info("Begin to validate the ip connect")
+            args.master_addr = ip_list[0]
+            for ip in ip_list[1:]:
+                completed_process = subprocess.run("ssh -o PasswordAuthentication=no {} ':'".format(ip), shell=True)
```

Comment:
Please use f-strings
```suggestion
                completed_process = subprocess.run(f"ssh -o PasswordAuthentication=no {ip} ':'", shell=True)
```